### PR TITLE
Macro for entering permutations in cycle notation

### LINKED ISSUE #420: Feature: nicer input syntax for permutations
This is an older idea but I am not sure I ever wrote it down, so here goes: it might be kinda nice to add a macro that makes it convenient to enter permutations in cycle notation, GAP style. This can be achieved using macros. With the [MacroTools.jl](https://github.com/FluxML/MacroTools.jl) package, it is particularly convenient:
```
julia> macro perm(ex)
         MacroTools.postwalk(ex) do x
           @capture(x, f_(xs__)) || return x
           :($f, ($(xs...),))
         end
       end
@perm (macro with 1 method)

julia> @perm (1,2)(3,4,5)(7,8,9)
(((1, 2), (3, 4, 5)), (7, 8, 9))
```
This works because Julia parses the expression `(1,2)(3,4,5)` as a function call, where `(1,2)` is the function and `(3,4,5)` are the arguments. 

With some tweaking, it should be possible to tweak this code to produce a GAP permutation. The easiest approach of course would be to convert those tuples into GAP permutations and then multiply them, but that's (a) not very efficient and (b) does not reject the case where the cycles overlap.

UPDATE: to clarify, this shouldn't just work for integers, but also nested expressions, e.g. this: `a=1;b=2; pi = @perm (a,f(a),b)(a+1,b*2)` (to give a rather arbitrary example)

### LINKED ISSUE #1189: Nicer syntax for entering permutations
Right now, you may see something like this in a REPL or notebook:
```
julia> G
<permutation group of size 645120 with 9 generators>

julia> gens(G)
9-element Vector{PermGroupElem}:
 (1,10)
 (2,11)
 (3,12)
 (4,13)
 (5,14)
 (6,8)
 (7,9)
 (1,2,3,4,5,6,7)(8,9,10,11,12,13,14)
 (1,2)(10,11)
```
But how can you enter this group again, given just this output? 

One idea would be to write a macro which can take a single permutation or a list and turn them into OSCAR permutation (we already had a similar idea for GAP.jl, see See also https://github.com/oscar-system/GAP.jl/issues/420)

So then you could perhaps write
```
gens = @perm 14 [
  (1,10)
 (2,11)
 (3,12)
 (4,13)
 (5,14)
 (6,8)
 (7,9)
 (1,2,3,4,5,6,7)(8,9,10,11,12,13,14)
 (1,2)(10,11)
]
```
and it would produce the list of permutations as elements of `sym(14)`. Alternatively the first argument `14` could be omitted (then just take the maximum of all occurring moved points), or could be an existing permutation group.

(Of course instead of returning a list of permutations, it could also directly return a permutation group. I just figured being able to deal with a list of permutations is more general).

In principle it is clear to me how to implement this, but I figured it would be good to first get some feedback, perhaps people have better ideas for the syntax of this, which I'd like to hear before working on that macro :-)

### PR DESCRIPTION
**Pull request by danielrademacher**
Resolves https://github.com/oscar-system/GAP.jl/issues/420
Resolves #1189

### CODE REVIEW

File: experimental/Permutations/Permutations.jl [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+#              (1,2,3,4,5,6,7)(8,9,10,11,12,13,14)
+#              (1,2)(10,11)
+#             ];
+# julia> G = Oscar.Permutations.permgroup(14,gens);
+# julia> Oscar.Permutations.size(G)
+# > 645120
+#
+function size(G::PermGroup)
+    return GAP.Globals.Size(G.X)
+end
```
Comment (thofma) [inline experimental/Permutations/Permutations.jl]: Hm, why is this added here?

File: experimental/Permutations/Permutations.jl [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+#              (5,14)
+#              (6,8)
+#              (7,9)
+#              (1,2,3,4,5,6,7)(8,9,10,11,12,13,14)
+#              (1,2)(10,11)
+#             ];
+# julia> G = Oscar.Permutations.permgroup(14,gens)
+# ><permutation group with 9 generators>
+#
+function permgroup(n::Int64,gens::Vector{PermGroupElem})
```
Comment (thofma) [inline experimental/Permutations/Permutations.jl]: Why is this added here? Should go into the corresponding file in `src/Groups`. @fingolfin Should this be a `PermGroup` constructor?

**Comment by thofma** [comment]
Would be good if there were some tests with invalid input, e.g. `@perm (-1, 1)` or `@perm 10 [(1 11)]`.

File: experimental/Experimental.jl [hunk 3]
```diff
@@ -6,6 +6,8 @@ include("InvariantTheory.jl")
 include("GITFans.jl")
 include("GModule.jl")
 
+include("Permutations/Permutations.jl")
```
Comment (fingolfin) [inline experimental/Experimental.jl]: I don't think we have to keep this in experimental. Just put it all into `src/Groups/perm.jl`?

File: experimental/Permutations/Permutations.jl [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+# it is possible to create a permutation as follow
+# pi = Oscar.Permutations.@perm (1,2,3)(4,5)(6,7,8)
+# > (1,2,3)(4,5)(6,7,8)
+# For this we use macros to modify the syntax tree of (1,2,3)(4,5)(6,7,8) such that
+# Julia can deal with the expression.
+
+module Permutations
+
+ using GAP
+ using Oscar
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: Scratch that and put it into `src/Groups/perm.jl` directly into the top module :-)

File: experimental/Permutations/Permutations.jl [hunk 5]
```diff
@@ -0,0 +1,136 @@
+# This file implements a new way to input permutations in Julia. For example
+# it is possible to create a permutation as follow
+# pi = Oscar.Permutations.@perm (1,2,3)(4,5)(6,7,8)
+# > (1,2,3)(4,5)(6,7,8)
+# For this we use macros to modify the syntax tree of (1,2,3)(4,5)(6,7,8) such that
+# Julia can deal with the expression.
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: This should be part of a docstring for the new macro

File: experimental/Permutations/Permutations.jl [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+# Macro to input a permutation as
+# julia> pi = Oscar.Permutations.@perm (1,2,3)(4,5)(6,7,8)
+# > (1,2,3)(4,5)(6,7,8)
+#
+macro perm(ex)
+    h = (ex).head
+    arg = (ex).args
+    res = []
+    
+    while h != :tuple
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: Needs some more error testing. I would actually recommend to do
```suggestion
    while h == :call
```
and then after the loop verify that the final expression really is a `:tuple`

File: experimental/Permutations/Permutations.jl [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+################################################################################
+# Macro to input a permutation as
+# julia> pi = Oscar.Permutations.@perm (1,2,3)(4,5)(6,7,8)
+# > (1,2,3)(4,5)(6,7,8)
+#
+macro perm(ex)
+    h = (ex).head
+    arg = (ex).args
+    res = []
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]:
```suggestion
    res = Expr[]
```
for type stability. Might not really matter here, but it's a good habit to learn about this sooner rather than later: if you write `[]` it will be a `Vector{Any}`, with my change a `Vector{Expr}` and that then allows for more optimization opportunities later on.

File: experimental/Permutations/Permutations.jl [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    
+    while h != :tuple
+        arg1 = arg[1]
+        insert!(res,1,Expr(:vect,arg[2:length(arg)]...))
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: The following two changes make it more idiomatic
```suggestion
        pushfirst!(res, Expr(:vect,arg[2:end]...))
```

File: experimental/Permutations/Permutations.jl [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        h = (arg1).head
+        arg = (arg1).args
+    end
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: This is a matter of taste I guess, but how about replacing `arg1` by just ex, i.e., write the code more like this:
```julia
  while ex.head == :call
     pushfirst!(res, Expr(:vect, ex.args[2:end]...))
     ex = ex.args[1]
  end
```

File: experimental/Permutations/Permutations.jl [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    
+    insert!(res,1,Expr(:vect,arg...))
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: Same as above and of course the type checking

File: experimental/Permutations/Permutations.jl [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+    return esc(:(Oscar.cperm($(res...))))
+end
+
+################################################################################
+# Macro to input a list of permutation which are generated as elements of
+# the symmetric_group(n)
+# Example:
+# julia> gens = Oscar.Permutations.@perm 14 [
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: This all should become a docstring

File: experimental/Permutations/Permutations.jl [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+# > Sym( [ 1 .. 14 ] )
+macro perm(n,gens)
+
+    s = symmetric_group(n)
+    ores = Vector{Expr}(undef,length(gens.args))
+    i = 1
+    for g in gens.args
+        h = g.head
+        arg = g.args
+        res = []
```
Comment (fingolfin) [inline experimental/Permutations/Permutations.jl]: Similar comments as above apply.

You could also factor out some of the common code into a helper function (not macro!) that is called by both macros.

I can write more but I'd rather first wait for you to address some of my other comments, then it'll be less confusing

File: experimental/Permutations/Permutations.jl (hunk 1 above)
Comment (fingolfin) [reply experimental/Permutations/Permutations.jl]: It should be removed and `order` be used instead

File: test/Experimental/Permutations-test.jl [hunk 13]
```diff
@@ -0,0 +1,42 @@
+@testset "Examples.Permutations" begin
+  p = Oscar.Permutations.@perm (1,2)(3,4)(5,6)
```
Comment (fingolfin) [inline test/Experimental/Permutations-test.jl]: Make sure to `export @perm` (and also any other macros), then you can just write
```suggestion
  p = @perm (1,2)(3,4)(5,6)
```

File: test/Experimental/Permutations-test.jl [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+  @test p  == cperm([1,2],[3,4],[5,6])
+  p = Oscar.Permutations.@perm (1,2)(3,4,5)(7,8,9)
+  @test p == cperm([1,2],[3,4,5],[7,8,9])
+
+  a=1;b=2;f=x->3*x + 2;
+  p = Oscar.Permutations.@perm (a,f(a),b)(a+1,b*2)
+  @test p == cperm([1,5,4,2])
+  
+  @test_throws ErrorException Oscar.Permutations.@perm (-1, 1)
```
Comment (fingolfin) [inline test/Experimental/Permutations-test.jl]: I'd add some more error tests. E.g. what happens if I do `@perm "bla"` or `@perm 1 + 1` ?

**Review by fingolfin — CHANGES_REQUESTED** [review]
Thank you!! Great work. Still some changes would be good

File: test/Groups/Permutations.jl [hunk 15]
```diff
@@ -0,0 +1,44 @@
+@testset "Examples.Permutations" begin
+  p = @perm (1,2)(3,4)(5,6)
+  @test p  == cperm([1,2],[3,4],[5,6])
```
Comment (fingolfin) [inline test/Groups/Permutations.jl]:
```suggestion
  @test p == cperm([1,2],[3,4],[5,6])
```

File: experimental/Permutations/Permutations.jl (hunk 7 above)
Comment (danielrademacher) [reply experimental/Permutations/Permutations.jl]: This throws an error
```
ERROR: LoadError: TypeError: in Expr, expected Symbol, got a value of type Vector{Any}
```

Comment (fingolfin) [reply experimental/Permutations/Permutations.jl]: OK, maybe I got the type wrong then. You could insert something like `println(typeof(res))` at the end of the macro to see what it typically is like.

Anyway, I think you should just forget about it, it's not worth it: this code is only executed during precompilation or in a REPL, and a few microseconds here or there won't matter

File: src/Groups/perm.jl [hunk 16]
```diff
@@ -387,3 +387,138 @@ function cycle_structure(g::PermGroupElem)
...
+# For this we use macros to modify the syntax tree of (1,2,3)(4,5)(6,7,8) such that
+# Julia can deal with the expression.
+
+
+################################################################################
+#
+#   perm
+#
+@doc Markdown.doc"""
+ @perm(ex)
```
Comment (fingolfin) [inline src/Groups/perm.jl]: First line of a docstring should be indented four spaces (to start a Markdown code block), and be separated from the following text by an empty line.
```suggestion
    @perm(ex)
```

File: src/Groups/perm.jl [hunk 17]
```diff
@@ -387,3 +387,138 @@ function cycle_structure(g::PermGroupElem)
...
+Macro to input a permutation as
+pi = @perm (1,2,3)(4,5)(6,7,8) to obtain
```
Comment (fingolfin) [inline src/Groups/perm.jl]: This is Markdown, and code should be marked as such, i.e.
```suggestion
`pi = @perm (1,2,3)(4,5)(6,7,8)` to obtain
```

File: src/Groups/perm.jl [hunk 18]
```diff
@@ -387,3 +387,138 @@ function cycle_structure(g::PermGroupElem)
...
+(1,2,3)(4,5)(6,7,8)
```
Comment (fingolfin) [inline src/Groups/perm.jl]: The user may not know what that means, though. Better perhaps to explain it in terms of the function `cperm`. Perhaps like this
```suggestion
the permutation `(1,2,3)(4,5)(6,7,8)`, that is, the output of
`cperm([1,2,3],[4,5],[6,7,8])`.
```

File: src/Groups/perm.jl [hunk 19]
```diff
@@ -387,3 +387,138 @@ function cycle_structure(g::PermGroupElem)
...
+# Examples (jldoctest)
+```
```
Comment (fingolfin) [inline src/Groups/perm.jl]:
````suggestion
# Examples
```jldoctest
````

File: src/Groups/perm.jl [hunk 20]
```diff
@@ -387,3 +387,138 @@ function cycle_structure(g::PermGroupElem)
...
+julia> @perm (1,2,3)(4,5)(6,7,8)
+(1,2,3)(4,5)(6,7,8)
+```
+"""
+macro perm(ex)
+    res = []
+   
+    while (ex).head == :call
```
Comment (fingolfin) [inline src/Groups/perm.jl]: No need to write `(ex)`, just change it to `ex` everywhere.
```suggestion
    while ex.head == :call
```

File: test/runtests.jl [hunk 21]
```diff
@@ -31,6 +31,7 @@ include("Groups/runtests.jl")
 include("Rings/NumberField.jl")
 include("Rings/FunctionField-test.jl")
 include("Rings/AbelianClosure.jl")
+include("Groups/Permutations.jl")
```
Comment (fingolfin) [inline test/runtests.jl]: This really should be loaded from `Groups/runtests.jl` instead

**Comment by danielrademacher** [comment]
I also added a check for 
```
@perm ()
```

But than an error occurs for cperm 


```
cperm([])
ERROR: MethodError: no method matching cperm(::Vector{Any})
Closest candidates are:
  cperm(::AbstractVector{T}...) where T<:Union{Integer, fmpz} at ~/.julia/dev/Oscar/src/Groups/perm.jl:214
  cperm(::PermGroup, ::AbstractVector{T}...) where T<:Union{Integer, fmpz} at ~/.julia/dev/Oscar/src/Groups/perm.jl:226
```

Is this intended?

File: src/Groups/perm.jl [hunk 22]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+julia> @perm (1,2,3)(4,5)(6,7,8)
+(1,2,3)(4,5)(6,7,8)
+```
+"""
+macro perm(ex)
+    res = []
+    
+    if typeof(ex) != Expr
+        error("Input is not a permutation.")
+    end
```
Comment (fingolfin) [inline src/Groups/perm.jl]: To check a type, use `isa` instead of comparing the value of `typeof`. Also, this kind of check could be written as follows, which is more idiomatic for Julia code:
```suggestion
    ex isa expr || error("Input is not a permutation expression")
```

File: src/Groups/perm.jl [hunk 23]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+    
+    if ex.head != :call && ex.head != :tuple
+        error("Input is not a permutation.")
+    end
+
```
Comment (fingolfin) [inline src/Groups/perm.jl]: This is redundant: if the condition is met, then the `while` loop is not executed, and we run into the check after it
```suggestion
```

File: src/Groups/perm.jl [hunk 24]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+   
+    while ex.head == :call
+        pushfirst!(res, Expr(:vect, ex.args[2:end]...))
+        ex = ex.args[1]
+        if typeof(ex) != Expr
+            error("Input is not a permutation.")
+        end
+        if ex.head != :call && ex.head != :tuple
+            error("Input is not a permutation.")
+        end
```
Comment (fingolfin) [inline src/Groups/perm.jl]: Redundant
```suggestion
```

File: src/Groups/perm.jl [hunk 25]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+    if ex.head != :call && ex.head != :tuple
+        error("Input is not a permutation.")
+    end
+   
+    while ex.head == :call
+        pushfirst!(res, Expr(:vect, ex.args[2:end]...))
+        ex = ex.args[1]
+        if typeof(ex) != Expr
+            error("Input is not a permutation.")
+        end
```
Comment (fingolfin) [inline src/Groups/perm.jl]:
```suggestion
        ex isa Expr || error("Input is not a permutation expression")
```

File: src/Groups/perm.jl [hunk 26]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+    
+    if typeof(ex) != Expr
+        error("Input is not a permutation.")
+    end
+    
+    if ex.head != :call && ex.head != :tuple
+        error("Input is not a permutation.")
+    end
+   
+    while ex.head == :call
```
Comment (fingolfin) [inline src/Groups/perm.jl]: To simplify the error handling further, how about doing this here:
```suggestion
    while ex isa Expr && ex.head == :call
```
Then a single `error` check after the while loop is enough

File: src/Groups/perm.jl [hunk 27]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+# Julia can deal with the expression.
+
+
+################################################################################
+#
+#   perm
+#
+@doc Markdown.doc"""
+    @perm(ex)
+Macro to input a permutation as
```
Comment (fingolfin) [inline src/Groups/perm.jl:431]: The empty line between the function signature and the description is important
```suggestion

Macro to input a permutation as
```

File: src/Groups/perm.jl [hunk 28]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+    return esc(:(Oscar.cperm($(res...))))
+end
+
+
+################################################################################
+#
+#   perm(n,gens)
+#
+@doc Markdown.doc"""
+    @perm(n,gens)
```
Comment (fingolfin) [inline src/Groups/perm.jl:454]:
```suggestion
    @perm(n,gens)

```

File: src/Groups/perm.jl [hunk 29]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+Macro to input a list of permutations which are generated as elements of
+the `symmetric_group(n)` with the function `cperm`.
+# Examples
+```jldoctest
+julia> gens = Oscar.Permutations.@perm 14 [
+               (1,10)
```
Comment (fingolfin) [inline src/Groups/perm.jl]:
```suggestion
              (1,10)
```

File: src/Groups/perm.jl [hunk 30]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+9-element Vector{PermGroupElem}:
+ (1,10)
+ (2,11)
+ (3,12)
+ (4,13)
+ (5,14)
+ (6,8)
+ (7,9)
+ (1,2,3,4,5,6,7)(8,9,10,11,12,13,14)
+ (1,2)(10,11)
```
Comment (fingolfin) [inline src/Groups/perm.jl:480]: Output should appear exactly as it appears in the REPL, which also means an empty line comes between the output and the next prompt.
```suggestion
 (1,2)(10,11)

```

File: src/Groups/perm.jl [hunk 31]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+julia> gens[1].parent
```
Comment (fingolfin) [inline src/Groups/perm.jl]: Never access struct members directly like that (unless you are implementing accessor functions). So:
```suggestion
julia> parent(gens[1])
```

File: src/Groups/perm.jl [hunk 32]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+Sym( [ 1 .. 14 ] )
+```
+"""
+macro perm(n,gens)
+
+    s = symmetric_group(n)
```
Comment (fingolfin) [inline src/Groups/perm.jl]: I don't think this is needed or used?
```suggestion
```

File: src/Groups/perm.jl [hunk 33]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+            if ex.head != :call && ex.head != :tuple
+                error("Input is not a permutation.")
+            end
+        end
+        
+        if ex.head != :tuple
+            error("Input is not a permutation.")
+        end
+        
+        pushfirst!(res, Expr(:vect,ex.args...))
```
Comment (fingolfin) [inline src/Groups/perm.jl]: This code is essentially equivalent to the body of the macro `@perm`. Let's share code! To do that, extract most of the body of the `macro perm` and move it into a helper function, but be sure to let it return `res`, not an expression derived from that. I.e.:
```julia
function _perm_helper(ex::Expr)
  ...
  return res
end
```
Then we can write
```julia
macro perm(ex)
   ex isa Expr || error("Input is not a permutation expression)
   res = _perm_helper(ex)
   return esc(:(Oscar.cperm($(res...))))
end
```
But also use it in this macro here.

File: src/Groups/perm.jl [hunk 34]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+
+        ores[i] = esc(:(Oscar.cperm(symmetric_group($n),$(res...))))
+        i = i + 1
```
Comment (fingolfin) [inline src/Groups/perm.jl]: You could just let `ores` start out as an empty vector and then do `push!(ores, ...)`, then you don't need `i`.

File: src/Groups/perm.jl [hunk 35]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+#
+#   perm(n,gens)
+#
+@doc Markdown.doc"""
+    @perm(n,gens)
+Macro to input a list of permutations which are generated as elements of
+the `symmetric_group(n)` with the function `cperm`.
+# Examples
+```jldoctest
+julia> gens = Oscar.Permutations.@perm 14 [
```
Comment (fingolfin) [inline src/Groups/perm.jl]: `Oscar.Permutations` does not exist
```suggestion
julia> gens = @perm 14 [
```

File: src/Groups/perm.jl [hunk 36]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+            end
+        end
+        
+        if ex.head != :tuple
+            error("Input is not a permutation.")
+        end
+        
+        pushfirst!(res, Expr(:vect,ex.args...))
+
+        ores[i] = esc(:(Oscar.cperm(symmetric_group($n),$(res...))))
```
Comment (fingolfin) [inline src/Groups/perm.jl]: This calls `symmetric_group` again and again, which isn't great, as we create a new group object each time. This could be improved by using `let` block; but I don't want to complicate things even further, so I suggest we first iron out everything else and get this merged, then we can worry about fine tuning.

Just for the record, here is what I'd do: I'd only collect the `res` values here:
```suggestion
        ores[i] = esc(:(  [$(res...)] ))
```

Then, the final `return` of this macro could look something like this:
```julia
return quote
   let g = symmetric_group($n)
       [ cperm(g, pi) for pi in [$(ores...)] ]
   end
end
```
Note: normally we'd have to worry that `g` and `pi` could clash with symbols defined by the programmer in the surrounding code, and referenced in the permutation expression.  This is where the `esc` function comes into play: we applied it to the expressions we got from the user, but *not* to the outer expression above. This leads to the Julia [macro hygiene](https://docs.julialang.org/en/v1/manual/metaprogramming/#Hygiene) functionality kicking in, which determines that `g` and `pi` (unlike `symmetric_group` and `cperm`) are not known variables, and replaces them by special generated names. You can check the effect of this via `macroexpand`:
```
julia> g = 7  # introduce a variable g and use it in the macro expression:
7

julia> macroexpand(Main, :( Oscar.@perm 14 [
                     (g,9)
                     (1,2)(10,11)
                    ]))
quote
    let var"#332#g" = Oscar.symmetric_group(14)
        [Oscar.cperm(var"#332#g", var"#333#pi"...) for var"#333#pi" = [[[g, 9]], [[1, 2], [10, 11]]]]
    end
end
```

File: src/Groups/perm.jl [hunk 37]
```diff
@@ -387,3 +387,168 @@ function cycle_structure(g::PermGroupElem)
...
+              (2,11)
+              (3,12)
+              (4,13)
+              (5,14)
+              (6,8)
+              (7,9)
+              (1,2,3,4,5,6,7)(8,9,10,11,12,13,14)
+              (1,2)(10,11)
+             ];
+julia> G = Oscar.Permutations.permgroup(14,gens)
```
Comment (fingolfin) [inline src/Groups/perm.jl]: I don't see why this is needed? Just write
```suggestion
julia> G = sub(symmetric_group(14),gens)[1]
```

File: test/Groups/Permutations.jl [hunk 38]
```diff
@@ -0,0 +1,44 @@
...
+  @test p == cperm([1,2],[3,4],[5,6])
+  p = @perm (1,2)(3,4,5)(7,8,9)
+  @test p == cperm([1,2],[3,4,5],[7,8,9])
+
+  a=1;b=2;f=x->3*x + 2;
+  p = @perm (a,f(a),b)(a+1,b*2)
+  @test p == cperm([1,5,4,2])
+  
+  @test_throws ErrorException @perm (-1, 1)
+  @test_throws LoadError @perm "bla"
```
Comment (fingolfin) [inline test/Groups/Permutations.jl]: Ahhh now I understand where your `LoadError` comes from: the problem is that `@perm` is evaluated right away, while reading the file, and so `@test_throws` has no chance of catching that error.

You need to defer the expansion of the macro `@perm` in order to test it adequately. This can be achieved via `eval` resp.  the `@eval` macro. Try this:
```suggestion
  @test_throws LoadError @eval @perm "bla"
```

**Review by fingolfin — APPROVED** [review]
Thanks!
